frontend/src/app/main/ui/workspace/sidebar/layers.cljs

```diff
@@ -84,101 +84,114 @@
-       (l/derived refs/workspace-local)))
-
 (mf/defc layer-item
   [{:keys [index item selected objects] :as props}]
   (let [id         (:id item)
-        selected?  (contains? selected id)
-        container? (or (cph/frame-shape? item)
-                       (cph/group-shape? item))
+        blocked?   (:blocked item)
+        hidden?    (:hidden item)
 
         disable-drag      (mf/use-state false)
         scroll-to-middle? (mf/use-var true)
+        expanded-iref     (mf/with-memo [id]
+                            (-> (l/in [:expanded id])
+                                (l/derived refs/workspace-local)))
 
-        expanded-iref (mf/use-memo
-                       (mf/deps id)
-                       (make-collapsed-iref id))
-
-        expanded? (mf/deref expanded-iref)
+        expanded?         (mf/deref expanded-iref)
+        selected?         (contains? selected id)
+        container?        (or (cph/frame-shape? item)
+                              (cph/group-shape? item))
 
         toggle-collapse
-        (fn [event]
-          (dom/stop-propagation event)
-          (if (and expanded? (kbd/shift? event))
-            (st/emit! (dwc/collapse-all))
-            (st/emit! (dwc/toggle-collapse id))))
+        (mf/use-fn
+         (mf/deps expanded?)
+         (fn [event]
+           (dom/stop-propagation event)
+           (if (and expanded? (kbd/shift? event))
+             (st/emit! (dwc/collapse-all))
+             (st/emit! (dwc/toggle-collapse id)))))
 
         toggle-blocking
-        (fn [event]
-          (dom/stop-propagation event)
-          (if (:blocked item)
-            (st/emit! (dw/update-shape-flags [id] {:blocked false}))
-            (st/emit! (dw/update-shape-flags [id] {:blocked true})
-                      (dw/deselect-shape id))))
+        (mf/use-fn
+         (mf/deps id blocked?)
+         (fn [event]
+           (dom/stop-propagation event)
+           (if blocked?
+             (st/emit! (dw/update-shape-flags [id] {:blocked false}))
+             (st/emit! (dw/update-shape-flags [id] {:blocked true})
+                       (dw/deselect-shape id)))))
 
         toggle-visibility
-        (fn [event]
-          (dom/stop-propagation event)
-          (if (:hidden item)
-            (st/emit! (dw/update-shape-flags [id] {:hidden false}))
-            (st/emit! (dw/update-shape-flags [id] {:hidden true}))))
+        (mf/use-fn
+         (mf/deps hidden?)
+         (fn [event]
+           (dom/stop-propagation event)
+           (if hidden?
+             (st/emit! (dw/update-shape-flags [id] {:hidden false}))
+             (st/emit! (dw/update-shape-flags [id] {:hidden true})))))
 
         select-shape
-        (fn [event]
-          (dom/prevent-default event)
-          (reset! scroll-to-middle? false)
-          (let [id (:id item)]
-            (cond
-              (kbd/shift? event)
-              (st/emit! (dw/shift-select-shapes id))
+        (mf/use-fn
+         (mf/deps id)
```

Comment:
Won't use-fn have the same result if we omit this expr?

Reply:
`use-fn` in this case is the React.useCallback hook, it allows having stable reference to the function callback instance upon all renders while the id not changes. This makes that on VDOM diffing it does not need to remove the old callback from the real dom and attach a new one. On normal code it can be really irrelevant, but in code that renders a lot and is in the "hot path" every millisecond counts and this helps on this.

Reply:
Thanks for giving details about performance, I see how memo is of use.:)
I was wondering, could we omit (mf/deps id), having it as
```
(mf/use-fn
  (fn [event]
    ...
```
to the same extent?

Reply:
The rational behind it is that we don't list `id` in other memos, since it does not change, so if we are to stay uniform perhaps here we'd have empty deps or, rather, absent deps

Reply:
The id dep can be omited if we ensure when layer-item component is always used with id as :key,  but this is a error prone asumption and in this case is better provide the hook with deps.